File: app/src/main/java/com/coloryr/facetrack/live2d/JniBridgeJava.java
/**
 * Copyright(c) Live2D Inc. All rights reserved.
 * <p>
 * Use of this source code is governed by the Live2D Open Software license
 * that can be found at https://www.live2d.com/eula/live2d-open-software-license-agreement_en.html.
 */

package com.coloryr.facetrack.live2d;

import android.annotation.SuppressLint;
import android.content.Context;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class JniBridgeJava {

    private static final String LIBRARY_NAME = "live2d";
    @SuppressLint("StaticFieldLeak")
    private static Context _context;

    static {
        System.loadLibrary(LIBRARY_NAME);
    }

    // Native -----------------------------------------------------------------

    public static native void nativeOnStart();

    public static native void nativeOnStop();

    public static native void nativeOnDestroy();

    public static native void nativeOnSurfaceCreated();

    public static native void nativeOnSurfaceChanged(int width, int height);

    public static native void nativeOnDrawFrame();

    public static native void nativeOnTouchesBegan(float pointX, float pointY);

    public static native void nativeOnTouchesEnded(float pointX, float pointY);

    public static native void nativeOnTouchesMoved(float pointX, float pointY);

    private static native void nativeStartMotion(byte[] group, int no, int priority);

    private static native void nativeStartExpressions(byte[] name);

    public static native int nativeGetExpressionSize();

    public static native int nativeGetMotionSize();

    public static native byte[] nativeGetExpression(int index);

    public static native byte[] nativeGetMotion(int index);

    public static native void nativeSetPosX(float x);

    public static native void nativeSetPosY(float y);

    public static native void nativeSetPos(float x, float y);

    public static native void nativeSetScale(float scale);

    public static native void nativeLoadModel(byte[] path, byte[] name);

    public static native void nativeEnableRandomMotion(boolean enable);

    public static native void nativeSetBreath(byte[] id);

    public static native void nativeGetCubismParams();

    public static native void nativeSetCubismParams(byte[] id, float value);

    public static native void nativeSetEyeBallX(byte[] id);

    public static native void nativeSetEyeBallY(byte[] id);

    public static native void nativeSetBodyAngleX(byte[] id);

    public static native void nativeSetAngleX(byte[] id);

    public static native void nativeSetAngleY(byte[] id);

    public static native void nativeSetAngleZ(byte[] id);

    public static native String[] nativeGetParamIds();

    public static native float[] nativeGetPartValues();

    public static native String[] nativeGetPartIds();

    public static native float[] nativeGetParamValues();

    public static native float[] nativeGetParamDefaultValues();

    public static native float[] nativeGetParamMinimumValues();

    public static native float[] nativeGetParamMaximumValues();

    public static native void nativeSetParamValue(byte[] index, float value);

    // Java -----------------------------------------------------------------

    private static final List<String> motions = new ArrayList<>();
    private static final List<String> expressions = new ArrayList<>();

    private static void resData() {
        int a = nativeGetMotionSize();
        for (int i = 0; i < a; i++) {
            String temp = new String(nativeGetMotion(i));
            motions.add(temp);
        }
        a = nativeGetExpressionSize();
        for (int i = 0; i < a; i++) {
            String temp = new String(nativeGetExpression(i));
            expressions.add(temp);
        }
    }

    public static List<String> getMotions() {
        if (motions.size() == 0) {
            resData();
        }
        return motions;
    }

    public static List<String> getExpressions() {
        if (expressions.size() == 0) {
            resData();
        }
        return expressions;
    }

    public static void ChangeModel() {
        motions.clear();
        expressions.clear();
    }

    public static void StartMotion(String name, int no, int priority) {
        if (motions.size() == 0) {
            resData();
        }
        if (motions.contains(name + '_' + no)) {
            nativeStartMotion(name.getBytes(StandardCharsets.UTF_8), no, priority);
        }
    }

    public static void StartExpressions(String name) {
        if (expressions.size() == 0) {
            resData();
        }
        if (expressions.contains(name)) {
            nativeStartExpressions(name.getBytes(StandardCharsets.UTF_8));
        }
    }

    public static void SetContext(Context context) {
        _context = context;
    }

    public static byte[] LoadFile(String filePath) {
        InputStream fileData = null;
        try {
            fileData = _context.getAssets().open(filePath);
            int fileSize = fileData.available();
            byte[] fileBuffer = new byte[fileSize];
            fileData.read(fileBuffer, 0, fileSize);
            return fileBuffer;
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                if (fileData != null) {
                    fileData.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static void LoadModel(String path, String name) {
        byte[] path1 = path.getBytes(StandardCharsets.UTF_8);
        byte[] name1 = name.getBytes(StandardCharsets.UTF_8);
        nativeLoadModel(path1, name1);
    }

    static float a = -30;
    static boolean up = true;

    public static void onUpdate(){
        JniBridgeJava.nativeSetParamValue("PARAM_ANGLE_X".getBytes(StandardCharsets.UTF_8), TrackSave.AngleX);
        JniBridgeJava.nativeSetParamValue("PARAM_ANGLE_Y".getBytes(StandardCharsets.UTF_8), TrackSave.AngleY);
        JniBridgeJava.nativeSetParamValue("PARAM_ANGLE_Z".getBytes(StandardCharsets.UTF_8), TrackSave.AngleZ);
        JniBridgeJava.nativeSetParamValue("PARAM_MOUTH_OPEN_Y".getBytes(StandardCharsets.UTF_8), TrackSave.MouthOpenY);
        JniBridgeJava.nativeSetParamValue("PARAM_EYE_BALL_X".getBytes(StandardCharsets.UTF_8), TrackSave.EyeBallX);
        JniBridgeJava.nativeSetParamValue("PARAM_EYE_BALL_Y".getBytes(StandardCharsets.UTF_8), TrackSave.EyeBallY);
        JniBridgeJava.nativeSetParamValue("PARAM_BODY_Z".getBytes(StandardCharsets.UTF_8), TrackSave.BodyZ);
        JniBridgeJava.nativeSetParamValue("PARAM_BODY_Y".getBytes(StandardCharsets.UTF_8), TrackSave.BodyY);
        JniBridgeJava.nativeSetParamValue("PARAM_EYE_L_OPEN".getBytes(StandardCharsets.UTF_8), TrackSave.EyeLOpen);
        JniBridgeJava.nativeSetParamValue("PARAM_EYE_R_OPEN".getBytes(StandardCharsets.UTF_8), TrackSave.EyeROpen);
    }

    public static void onLoadModel(String name) {
        if (name.equals("shizuku.model3.json")) {
            JniBridgeJava.nativeSetBreath("PARAM_BREATH".getBytes(StandardCharsets.UTF_8));
            JniBridgeJava.nativeEnableRandomMotion(false);
        } else {
            JniBridgeJava.nativeEnableRandomMotion(true);
        }
    }

    public static CubismPart[] getCubismParts() {
        String[] list = JniBridgeJava.nativeGetPartIds();
        if (list == null) {
            return null;
        }

        CubismPart[] list1 = new CubismPart[list.length];
        float[] values = nativeGetPartValues();

        for (int a = 0; a < list.length; a++) {
            list1[a] = new CubismPart();
            list1[a].id = list[a];
            list1[a].opacities = values[a];
        }
        return list1;
    }

    public static CubismParam[] getCubismParams() {
        String[] list = JniBridgeJava.nativeGetParamIds();
        if (list == null) {
            return null;
        }

        CubismParam[] list1 = new CubismParam[list.length];
        float[] values = nativeGetParamValues();
        float[] defaultValues = nativeGetParamDefaultValues();
        float[] minValues = nativeGetParamMinimumValues();
        float[] maxValues = nativeGetParamMaximumValues();

        for (int a = 0; a < list.length; a++) {
            list1[a] = new CubismParam();
            list1[a].id = list[a];
            list1[a].value = values[a];
            list1[a].defaultValue = defaultValues[a];
            list1[a].minimumValue = minValues[a];
            list1[a].maximumValue = maxValues[a];
        }
        return list1;
    }
}

File: L2d_App/app/src/main/java/com/coloryr/facetrack/live2d/JniBridgeJava.kt
/**
 * Copyright(c) Live2D Inc. All rights reserved.
 *
 *
 * Use of this source code is governed by the Live2D Open Software license
 * that can be found at https://www.live2d.com/eula/live2d-open-software-license-agreement_en.html.
 */
package com.coloryr.facetrack.live2d

import android.annotation.SuppressLint
import android.content.Context
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

@SuppressLint("StaticFieldLeak")
object JniBridgeJava {
    private const val LIBRARY_NAME = "live2d"

    @SuppressLint("StaticFieldLeak")
    private var _context: Context? = null

    init {
        System.loadLibrary(LIBRARY_NAME)
    }

    // Native -----------------------------------------------------------------
    external fun nativeOnStart()
    external fun nativeOnStop()
    external fun nativeOnDestroy()
    external fun nativeOnSurfaceCreated()
    external fun nativeOnSurfaceChanged(width: Int, height: Int)
    external fun nativeOnDrawFrame()
    external fun nativeOnTouchesBegan(pointX: Float, pointY: Float)
    external fun nativeOnTouchesEnded(pointX: Float, pointY: Float)
    external fun nativeOnTouchesMoved(pointX: Float, pointY: Float)
    private external fun nativeStartMotion(group: ByteArray, no: Int, priority: Int)
    private external fun nativeStartExpressions(name: ByteArray)
    external fun nativeGetExpressionSize(): Int
    external fun nativeGetMotionSize(): Int
    external fun nativeGetExpression(index: Int): ByteArray?
    external fun nativeGetMotion(index: Int): ByteArray?
    external fun nativeSetPosX(x: Float)
    external fun nativeSetPosY(y: Float)
    external fun nativeSetPos(x: Float, y: Float)
    external fun nativeSetScale(scale: Float)
    external fun nativeLoadModel(path: ByteArray?, name: ByteArray?)
    external fun nativeEnableRandomMotion(enable: Boolean)
    external fun nativeSetBreath(id: ByteArray?)
    external fun nativeGetCubismParams()
    external fun nativeSetCubismParams(id: ByteArray?, value: Float)
    external fun nativeSetEyeBallX(id: ByteArray?)
    external fun nativeSetEyeBallY(id: ByteArray?)
    external fun nativeSetBodyAngleX(id: ByteArray?)
    external fun nativeSetAngleX(id: ByteArray?)
    external fun nativeSetAngleY(id: ByteArray?)
    external fun nativeSetAngleZ(id: ByteArray?)
    external fun nativeGetParamIds(): Array<String>?
    external fun nativeGetPartValues(): FloatArray
    external fun nativeGetPartIds(): Array<String>?
    external fun nativeGetParamValues(): FloatArray
    external fun nativeGetParamDefaultValues(): FloatArray
    external fun nativeGetParamMinimumValues(): FloatArray
    external fun nativeGetParamMaximumValues(): FloatArray
    external fun nativeSetParamValue(index: ByteArray?, value: Float)

    // Java -----------------------------------------------------------------
    var isLoad = false
        private set
    var name: String? = null
        private set
    private val motions: MutableList<String> = ArrayList()
    private val expressions: MutableList<String> = ArrayList()
    private fun resData() {
        var a = nativeGetMotionSize()
        for (i in 0 until a) {
            val temp = String(nativeGetMotion(i)!!)
            motions.add(temp)
        }
        a = nativeGetExpressionSize()
        for (i in 0 until a) {
            val temp = String(nativeGetExpression(i)!!)
            expressions.add(temp)
        }
    }

    fun getMotions(): List<String> {
        if (motions.size == 0) {
            resData()
        }
        return motions
    }

    fun getExpressions(): List<String> {
        if (expressions.size == 0) {
            resData()
        }
        return expressions
    }

    fun ChangeModel() {
        motions.clear()
        expressions.clear()
    }

    fun StartMotion(name: String, no: Int, priority: Int) {
        if (motions.size == 0) {
            resData()
        }
        if (motions.contains(name + '_' + no)) {
            nativeStartMotion(name.toByteArray(StandardCharsets.UTF_8), no, priority)
        }
    }

    fun StartExpressions(name: String) {
        if (expressions.size == 0) {
            resData()
        }
        if (expressions.contains(name)) {
            nativeStartExpressions(name.toByteArray(StandardCharsets.UTF_8))
        }
    }

    fun SetContext(context: Context?) {
        _context = context
    }

    fun LoadFile(filePath: String?): ByteArray? {
        var fileData: InputStream? = null
        return try {
            fileData = _context!!.assets.open(filePath!!)
            val fileSize = fileData.available()
            val fileBuffer = ByteArray(fileSize)
            fileData.read(fileBuffer, 0, fileSize)
            fileBuffer
        } catch (e: IOException) {
            e.printStackTrace()
            null
        } finally {
            try {
                fileData?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun LoadModel(path: String, name: String) {
        val path1 = path.toByteArray(StandardCharsets.UTF_8)
        val name1 = name.toByteArray(StandardCharsets.UTF_8)
        nativeLoadModel(path1, name1)
        isLoad = true
    }

    fun onUpdate() {
        nativeSetParamValue("PARAM_ANGLE_X".toByteArray(StandardCharsets.UTF_8), TrackSave.AngleX)
        nativeSetParamValue("PARAM_ANGLE_Y".toByteArray(StandardCharsets.UTF_8), TrackSave.AngleY)
        nativeSetParamValue("PARAM_ANGLE_Z".toByteArray(StandardCharsets.UTF_8), TrackSave.AngleZ)
        nativeSetParamValue("PARAM_MOUTH_OPEN_Y".toByteArray(StandardCharsets.UTF_8), TrackSave.MouthOpenY)
        nativeSetParamValue("PARAM_EYE_BALL_X".toByteArray(StandardCharsets.UTF_8), TrackSave.EyeBallX)
        nativeSetParamValue("PARAM_EYE_BALL_Y".toByteArray(StandardCharsets.UTF_8), TrackSave.EyeBallY)
        nativeSetParamValue("PARAM_BODY_Z".toByteArray(StandardCharsets.UTF_8), TrackSave.BodyZ)
        nativeSetParamValue("PARAM_BODY_Y".toByteArray(StandardCharsets.UTF_8), TrackSave.BodyY)
        nativeSetParamValue("PARAM_EYE_L_OPEN".toByteArray(StandardCharsets.UTF_8), TrackSave.EyeLOpen)
        nativeSetParamValue("PARAM_EYE_R_OPEN".toByteArray(StandardCharsets.UTF_8), TrackSave.EyeROpen)
    }

    fun onLoadModel(name: String) {
        JniBridgeJava.name = name
        if (name == "shizuku.model3.json") {
            nativeSetBreath("PARAM_BREATH".toByteArray(StandardCharsets.UTF_8))
            nativeEnableRandomMotion(false)
        } else {
            nativeEnableRandomMotion(true)
        }
    }

    val cubismParts: Array<CubismPart?>?
        get() {
            val list = nativeGetPartIds() ?: return null
            val list1 = arrayOfNulls<CubismPart>(list.size)
            val values = nativeGetPartValues()
            for (a in list.indices) {
                list1[a] = CubismPart()
                list1[a]!!.id = list[a]
                list1[a]!!.opacities = values[a]
            }
            return list1
        }
    val cubismParams: Array<CubismParam?>?
        get() {
            val list = nativeGetParamIds() ?: return null
            val list1 = arrayOfNulls<CubismParam>(list.size)
            val values = nativeGetParamValues()
            val defaultValues = nativeGetParamDefaultValues()
            val minValues = nativeGetParamMinimumValues()
            val maxValues = nativeGetParamMaximumValues()
            for (a in list.indices) {
                list1[a] = CubismParam()
                list1[a]!!.id = list[a]
                list1[a]!!.value = values[a]
                list1[a]!!.defaultValue = defaultValues[a]
                list1[a]!!.minimumValue = minValues[a]
                list1[a]!!.maximumValue = maxValues[a]
            }
            return list1
        }
}